src/Avalonia.Controls/ItemsControl.cs

```diff
@@ -502,25 +502,34 @@ private void UpdatePseudoClasses(int itemCount)
             IInputElement? from,
             bool wrap)
         {
-            IInputElement? result;
-            var c = from;
-
-            do
+            for(;;)
             {
-                result = container.GetControl(direction, c, wrap);
+                var result = container.GetControl(direction, from, wrap);
+
+                if (result is null || result == from)
+                {
+                    return null;
+                }
 
-                if (result != null &&
-                    result.Focusable &&
+                if (result.Focusable &&
                     result.IsEffectivelyEnabled &&
                     result.IsEffectivelyVisible)
                 {
                     return result;
                 }
 
-                c = result;
-            } while (c != null && c != from && direction != NavigationDirection.First && direction != NavigationDirection.Last);
+                direction = direction switch
+                {
+                    //We did not find an enabled first item. Move downwards until we find one.
+                    NavigationDirection.First => NavigationDirection.Down,
+
+                    //We did not find an enabled last item. Move upwards until we find one.
+                    NavigationDirection.Last => NavigationDirection.Up,
+                    _ => direction
+                };
```

Comment:
What will happen if all items are disabled, won't it be trapped in the cycle? Minimal test might be useful.

Reply:
Good spot. This was actually broken. I disabled wrapping in that case now so it should terminate correctly.

Reply:
Also added some tests.